src/v/pandaproxy/test/sharded_client_cache.cc

```diff
@@ -0,0 +1,75 @@
+// Copyright 2022 Redpanda Data, Inc.
+//
+// Use of this software is governed by the Business Source License
+// included in the file licenses/BSL.md
+//
+// As of the Change Date specified in that file, in accordance with
+// the Business Source License, use of this software will be governed
+// by the Apache License, Version 2.0
+
+#include "pandaproxy/sharded_client_cache.h"
+
+#include "kafka/client/configuration.h"
+
+#include <seastar/core/sleep.hh>
+#include <seastar/testing/thread_test_case.hh>
+#include <seastar/util/defer.hh>
+
+#include <boost/test/unit_test.hpp>
+
+#include <chrono>
+
+using namespace std::chrono_literals;
+
+namespace pp = pandaproxy;
+
+SEASTAR_THREAD_TEST_CASE(test_sharded_client_cache) {
+    pp::credential_t user = pp::default_user();
+    size_t cache_size = 10;
+
+    kafka::client::configuration cfg;
+    cfg.sasl_mechanism.set_value(ss::sstring{"SCRAM-SHA-256"});
+    cfg.scram_username.set_value(user.first);
+    cfg.scram_password.set_value(user.second);
+    BOOST_REQUIRE_EQUAL(cfg.scram_username.value(), user.first);
+    BOOST_REQUIRE_EQUAL(cfg.scram_password.value(), user.second);
+
+    pp::sharded_client_cache client_cache;
+    client_cache
+      .start(
+        ss::default_smp_service_group(),
+        to_yaml(cfg, config::redact_secrets::no),
+        cache_size,
+        true)
+      .get();
+    auto stop_client_cache = ss::defer(
+      [&client_cache]() { client_cache.stop().get(); });
+
+    // First client fetch will test the creation path
+    client_cache.fetch_client(user)
+      .then([&client_cache, &user](pp::client_ptr client1) {
+          // Second client fetch will test the reuse path
+          return client_cache.fetch_client(user).then(
+            [client1](pp::client_ptr client2) {
+                BOOST_TEST(
+                  client1->creation_time() == client2->creation_time());
+            });
+      })
+      .get();
+
+    // Wait a bit more than 30s for the client to become stale
+    ss::sleep(ss::lowres_clock::duration(31s)).get();
```

Comment:
This is a long time to wait in a test, how does a client become stale? Can that be configurable?